lib/puppet/util/windows/api_types.rb

```diff
@@ -53,11 +53,18 @@ def read_handle
     alias_method :read_word,  :read_uint16
     alias_method :read_array_of_wchar, :read_array_of_uint16
 
-    def read_wide_string(char_length, dst_encoding = Encoding::UTF_8, encode_options = {})
+    def read_wide_string(char_length, dst_encoding = Encoding::UTF_8, strip = false, encode_options = {})
       # char_length is number of wide chars (typically excluding NULLs), *not* bytes
       str = get_bytes(0, char_length * 2).force_encoding('UTF-16LE')
+
+      if strip
+        i = str.index("\u0000".encode('UTF-16LE'))
+        str = str[0..(i-1)] if i
+        str = "" if i == 0
```

Comment:
This is needed because we want to drop any `\u0000` event if the string starts with it.

Reply:
Using the `String[start, length]` form of substring instead of `String[Range]` handles that correctly:

```
irb(main):049:0> term = "\u0000".encode('UTF-16LE')
irb(main):050:0> str = "\u0000bar".encode('UTF-16LE')
irb(main):051:0> str.index(term)
=> 0
irb(main):052:0> str[0, 0]
=> ""
```

As does the i>0 case:

```
irb(main):053:0> str = "foo\u0000bar".encode('UTF-16LE')
irb(main):054:0> str.index(term)
=> 3
irb(main):055:0> str[0, 3]
=> "foo"
```